res/controllers/Gemini-G4V-scripts.js

```diff
@@ -0,0 +1,1802 @@
+        }
+
+        // Ignores off event
+        if (value === 0x00) {
+            return;
+        }
+
+        // Calculates the decks
+        var oldDeck = this.aDecks[channel-1];
+        var newDeck = (oldDeck === channel ? channel+1 : channel);
+
+        this.decks[oldDeck].activate(0);
+        this.decks[newDeck].activate(1);
+    };
+
+    // Browse button 0x93/0x11
+    // Simple: Loads track on preview
+    this.evBtnBrowse = function(channel, midino, value, status, group) {
+        if (g4v.debug) {
+            print("Controller.evBtnBrowse - channel:" + channel + " midino:0x" + midino.toString(16) + " value:0x" + value.toString(16) + " status:0x" + status.toString(16) + " group:" + group);
+        }
+
+        // Ignores off event
+        if (value === 0x00) {
+            return;
+        }
+
+        if (engine.getValue("[PreviewDeck1]", "play", 1)) {
+            engine.setValue("[PreviewDeck1]", "play", 0);
+            engine.setValue("[PreviewDeck]", "show_previewdeck", 0);
+        } else {
+            engine.setValue("[PreviewDeck]", "show_previewdeck", 1);
+            engine.setValue("[PreviewDeck1]", "LoadSelectedTrack", 1);
+            //dealy needed to
+            engine.beginTimer(100, "engine.setValue(\"[PreviewDeck1]\",\"play\", 1)", true);
+        }
+    };
+
+    // Browse button 0x93/0x12
+    // Simple: Changes browse mode
+    this.evBtnBack = function(channel, midino, value, status, group) {
+        if (g4v.debug) {
+            print("Controller.evBtnBack - channel:" + channel + " midino:0x" + midino.toString(16) + " value:0x" + value.toString(16) + " status:0x" + status.toString(16) + " group:" + group);
+        }
+
+        if (value === 0x00) {
+            return;
+        }
+
+        this.browseMode = (value === 0x7f);
+
+        engine.setValue("[Library]", "MoveFocusForward", 1);
+    };
+
+    // Browse Encode 0xb3/0x1e
+    // Simple: Browses library
+    // Shit: Browses lists
+    this.evEncBrowse = function(channel, midino, value, status, group) {
+        if (g4v.debug) {
+            print("Controller.evEncBrowse - channel:" + channel + " midino:0x" + midino.toString(16) + " value:0x" + value.toString(16) + " status:0x" + status.toString(16) + " group:" + group);
+        }
+
+        // If playing preview, beatjump. If not, scroll the Library
+        if (engine.getValue("[PreviewDeck1]", "play")) {
+            engine.setValue("[PreviewDeck1]", (value === 0x41 ? "beatjump_4_forward" : "beatjump_4_backward"), 1);
+        } else {
+            if (g4v.shift) {
+                engine.setValue("[Library]", (value === 0x41 ? "MoveRight" : "MoveLeft"), 1);
+            } else {
+                engine.setValue("[Library]", (value === 0x41 ? "MoveUp" : "MoveDown"), 1);
+            }
+        }
+    };
+};
+
+// Creates Object for Mixxx
+try {
+    var g4v = new MyController();
+} catch (e) {
+    print("Exception creating G4V object: "+e);
+}
```

Comment:
Please don't do this. That way the error message will only be printed on the terminal (and only if `--controllerDebug` was used) instead of displaying an error popup. The user will be surprised if the controller is not working without warning.

Reply:
Ok. How can I show a meaningful message to the user?

Reply:
Just don't wrap it in a try catch block. It will still open a popup that loading the controller mapping failed, including a backtrace.

Reply:
You can test this by throwing an exception in the constructor.